Modules/_pickle.c

```diff
@@ -6295,11 +6297,10 @@ load_mark(UnpicklerObject *self)
             return -1;
         }
 
-        if (self->marks == NULL)
-            self->marks = PyMem_NEW(Py_ssize_t, alloc);
-        else
-            PyMem_RESIZE(self->marks, Py_ssize_t, alloc);
+        Py_ssize_t *marks_old = self->marks;
+        PyMem_RESIZE(self->marks, Py_ssize_t, alloc);
         if (self->marks == NULL) {
+            PyMem_FREE(marks_old);
```

Comment:
@serhiy-storchaka 
Might make sense to not change `self->marks` at all instead of `PyMem_FREE()` when `PyMem_RESIZE()` fails?
It was implemented this way in 2.7: https://github.com/python/cpython/blob/00aebabc7139741fadfe877372c733a2160c7dbd/Modules/cPickle.c#L4661-L4669

Reply:
Agree, it makes sense.